### PR TITLE
Integration testing

### PR DESCRIPTION
**Pull request by dcposch**
**I've added a system to run WebTorrent Desktop, click around, wait for the UI to update, take screenshots, and make sure those screenshots match what the UI should look like.**

It's fun to watch the tests run...

This is the first step to fixing #765

#### Linting

`npm test`

#### Integration testing

`npm run integration-test`

#### Creating an integration test

See `test/index.js` for two examples.

If you add a new `setup.screenshotCreateOrCompare`, the first time you run tests afterward, missing screenshots will be created under `test/screenshots`. After that, existing screenshots are compared.

### I've added two basic tests to start out

- [x] Run the app. Make sure it starts and that the default torrents look right
- [x] Run the app with the Downloads folder missing. Check for the error message, then click on the Preferences link in the error message.

We need many more tests.

Time to automate the testing we used to do by hand: https://github.com/feross/webtorrent-desktop/pull/865#issuecomment-243949030

### CODE REVIEW

**Comment by dcposch** [comment]
@feross @mathiasvr @Flet @noamokman

**Comment by Goldob** [comment]
Wouldn't it be better to check contents with WebdriverIO instead of taking screenshots? These tests will fail each time a small change in UI appears. At least maybe the 'faulty' screenshot should be saved? This way we could use it for feedback or even replace the 'right' one with it if the change is intentional (this workflow still doesn't seem right thou).

File: test/index.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    .then(() => setup.wait())
+    .then(() => setup.screenshotCreateOrCompare(app, t, 'torrent-list-basic'))
+    .then(() => setup.endTest(app, t),
+          (err) => setup.endTest(app, t, err || 'error'))
+})
+
+test('show download path missing', function (t) {
+  fs.removeSync(DOWNLOAD_DIR)
+
+  t.timeoutAfter(10e3)
```
Comment (Goldob) [inline test/index.js]: Maybe it's just me, but this not quite enough for app initialization and all the tests.

**Comment by dcposch** [comment]
@Goldob I don't think so.
* Ad-hoc checking makes the tests a lot more work to write
* Even diffing the whole HTML is not as thorough as screenshot diffing. For example, it wouldn't catch an bug where hitting ESC from a video doesn't correctly restore window size.
* Chrome's own integration tests use screenshot diffing iirc
* Small UI changes will break a few tests, but the fix is as easy as deleting the offending screenshots and running the tests, which will recreate them with the new look.
* The resulting Github PR will then show, pixel by pixel, the exact UI changes that were made! This is pretty sweet IMO. https://github.com/blog/817-behold-image-view-modes

**Comment by Goldob** [comment]
@dcposch  Okay, now I'm convinced. Thanks for the explanation.

It makes a lot more sense with this GitHub feature.

File: test/index.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+const DOWNLOAD_DIR = path.join(setup.TEST_DATA_DIR, 'Downloads')
+fs.mkdirpSync(DOWNLOAD_DIR)
+
+test.onFinish(function () {
+  console.log('Removing test dir...')
+  fs.removeSync(setup.TEST_DATA_DIR)
+})
+
+test('app runs', function (t) {
+  t.timeoutAfter(10e3)
```
Comment (Flet) [inline test/index.js]: Huh, today I learned scientific notation is supported :)
10e3 = 10000

File: test/setup.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    path: path.join(__dirname, '..', 'node_modules', '.bin',
+      'electron' + (process.platform === 'win32' ? '.cmd' : '')),
+    args: [path.join(__dirname, '..')],
+    env: {NODE_ENV: 'test'}
+  })
+}
+
+// Starts the app, waits for it to load, returns a promise
+function waitForLoad (app, t) {
+  return app.start().then(function () {
```
Comment (Flet) [inline test/setup.js]: Should the saved preferences be reset at every load for consistency?

**Comment by Flet** [comment]
Very cool @dcposch! 👍

File: src/config.js [hunk 4]
```diff
@@ -89,22 +100,31 @@ function getConfigPath () {
...
+    return getPath('downloads')
   }
+}
 
-  const electron = require('electron')
+function getPath (key) {
+  if (process.type === 'renderer') {
+    return electron.remote.app.getPath(key)
+  } else {
+    electron.app.getPath(key)
```
Comment (noamokman) [inline src/config.js]: @dcposch missing return here

File: test/index.js (hunk 2 above)
Comment (dcposch) [inline test/index.js]: Yeah. 10e3 milliseconds = 10 seconds

File: test/setup.js (hunk 3 above)
Comment (dcposch) [inline test/setup.js]: I added `wipeTestDataDir` to delete everything. Not just saved prefs but all of config.json + torrents + torrent data. Clean slate.

File: src/config.js (hunk 4 above)
Comment (dcposch) [inline src/config.js]: fixed

**Comment by dcposch** [comment]
> Or at least maybe the 'faulty' screenshot should be saved? 
@Goldob yeah, good call. Done

**Comment by feross** [comment]
LGTM!

**Pull request merged**